libs/filer/files_client.go

```diff
@@ -165,40 +212,138 @@ func (w *FilesClient) Delete(ctx context.Context, name string, mode ...DeleteMod
+	var aerr *apierr.APIError
+	if errors.As(err, &aerr) && aerr.StatusCode == http.StatusNotFound {
 		return FileDoesNotExistError{absPath}
 	}
+	return err
+}
 
-	// This API returns 409 if the underlying path is a directory.
-	if aerr.StatusCode == http.StatusConflict {
-		return DirectoryNotEmptyError{absPath}
+func (w *FilesClient) recursiveDelete(ctx context.Context, name string) error {
+	filerFS := NewFS(ctx, w)
+	dirsToDelete := make([]string, 0)
+	callback := func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Files API does not allowing deleting non-empty directories. We instead
+		// collect the directories to delete and delete them once all the files have
+		// been deleted.
+		if d.IsDir() {
+			dirsToDelete = append(dirsToDelete, path)
+			return nil
+		}
+
+		return w.delete(ctx, path)
 	}
 
-	return err
+	// Walk the directory and delete all the files.
+	err := fs.WalkDir(filerFS, name, callback)
+	if err != nil {
+		return err
+	}
+
+	// Delete the directories in reverse order to ensure that the parent
+	// directories are deleted after the children. This is possible because
+	// fs.WalkDir walks the directories in lexicographical order.
+	for i := len(dirsToDelete) - 1; i >= 0; i-- {
+		err := w.delete(ctx, dirsToDelete[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *FilesClient) Delete(ctx context.Context, name string, mode ...DeleteMode) error {
+	if slices.Contains(mode, DeleteRecursively) {
+		return w.recursiveDelete(ctx, name)
+	}
+	return w.delete(ctx, name)
 }
 
 func (w *FilesClient) ReadDir(ctx context.Context, name string) ([]fs.DirEntry, error) {
-	return nil, filesNotImplementedError("ReadDir")
+	absPath, err := w.root.Join(name)
+	if err != nil {
+		return nil, err
+	}
+
+	iter := w.workspaceClient.Files.ListDirectoryContents(ctx, files.ListDirectoryContentsRequest{
+		DirectoryPath: absPath,
+	})
+
+	files, err := listing.ToSlice(ctx, iter)
+	var apierr *apierr.APIError
+
+	// This API returns a 404 if the specified path does not exist.
+	if errors.As(err, &apierr) && apierr.StatusCode == http.StatusNotFound {
+		return nil, NoSuchDirectoryError{absPath}
+	}
+	// This API returns 409 if the underlying path is a file.
+	if errors.As(err, &apierr) && apierr.StatusCode == http.StatusConflict {
+		return nil, NotADirectory{absPath}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	entries := make([]fs.DirEntry, len(files))
+	for i, file := range files {
+		entries[i] = filesApiDirEntry{
+			i: filesApiFileInfo{
+				absPath:      file.Path,
+				isDir:        file.IsDirectory,
+				fileSize:     file.FileSize,
+				lastModified: file.LastModified,
+			},
+		}
+	}
+
+	// Sort by name for parity with os.ReadDir.
+	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
+	return entries, nil
 }
 
 func (w *FilesClient) Mkdir(ctx context.Context, name string) error {
-	// Directories are created implicitly.
```

Comment:
This change allows empty directories to be created, something that was not possible before.